kotlin/internal/jvm/compile.bzl

```diff
@@ -359,6 +380,198 @@ def _run_merge_jdeps_action(ctx, toolchains, jdeps, outputs, deps):
+    if srcs.kt:
+        args.add("-source-roots=" + ":".join([_to_path(f) for f in srcs.kt]))
+    if srcs.java:
+        args.add("-java-source-roots=" + ":".join([_to_path(f) for f in srcs.java]))
+
+    # Required: Project structure
+    args.add("-project-base-dir=" + project_base_dir.path)
+    args.add("-output-base-dir=" + output_base_dir.path)
+    args.add("-caches-dir=" + caches_dir.path)
+
+    # Required: Output directories
+    args.add("-kotlin-output-dir=" + kotlin_out_dir.path)
+    args.add("-java-output-dir=" + java_out_dir.path)
+    args.add("-class-output-dir=" + class_out_dir.path)
+    args.add("-resource-output-dir=" + resource_out_dir.path)
+
+    # Libraries (classpath for compilation) - colon-separated
+    # Use full runtime jars (not ABI jars) because annotation processors need complete type information
+    classpath_files = depset(
+        transitive = [
+            _java_infos_to_compile_jars([compile_deps]),
+            transitive_runtime_jars,
+        ],
+    ).to_list()
+
+    if classpath_files:
+        args.add("-libraries=" + ":".join([_to_path(f) for f in classpath_files]))
+
+    # JDK home
+    args.add("-jdk-home=" + str(toolchains.java_runtime.java_home))
+
+    # Processor JARs (passed as positional arguments at the end, space-separated)
+    args.add_all(processor_jars, map_each = _to_path)
+
+    # Run KSP2 standalone
+    # Note: KSP2 standalone does NOT support Bazel persistent worker protocol
+    # (it doesn't understand --persistent_worker flag), so we run it as a regular action
+    ctx.actions.run(
+        executable = ctx.executable._ksp2_standalone,
+        arguments = [args],
+        inputs = depset(
+            srcs.kt + srcs.java + processor_jars,
+            transitive = [_java_infos_to_compile_jars([compile_deps])],
+        ),
+        outputs = [
+            kotlin_out_dir,
+            java_out_dir,
+            class_out_dir,
+            resource_out_dir,
+            project_base_dir,
+            output_base_dir,
+            caches_dir,
+        ],
+        mnemonic = "KotlinKsp2Standalone",
+        progress_message = "Running KSP2 processors for %{label}",
+    )
+
+    # Package generated sources into srcjars
+    kotlin_srcjar = ctx.actions.declare_file(
+        ctx.label.name + "-ksp2-kotlin-generated.srcjar",
+    )
+    java_srcjar = ctx.actions.declare_file(
+        ctx.label.name + "-ksp2-java-generated.srcjar",
+    )
+
+    _create_srcjar(ctx, kotlin_out_dir, kotlin_srcjar)
+    _create_srcjar(ctx, java_out_dir, java_srcjar)
+
+    # Package generated classes and resources into a jar
+    # This includes:
+    # - Classes generated directly by KSP processors (not compiled from sources)
+    # - Resources like META-INF/services, META-INF/proguard, etc.
+    generated_class_jar = ctx.actions.declare_file(
+        ctx.label.name + "-ksp2-generated-classes.jar",
+    )
+
+    # Merge class_out_dir and resource_out_dir into one jar
```

Comment:
Instead of `run_shell`, move the execution logic to a worker binary. 

Amoung other things, it makes unittesting easier -- and removes an action. They add up.